### ISSUE TITLE
(Help) Appending custom data to stream / augmenting question-answer pairs

### ISSUE THREAD

**Issue by dnsosebee** [vercel-labs/ai#350]
Wondering what the best solution is for appending custom extra data to a Q/A pair?

I have the following server-side code that appends suggested URLs to the stream:
```typescript
    // 10. Generate the stream that includes URL data
    const urlDataString = `data: ${JSON.stringify(urlData.slice(0, 9))}\n\n`;
    const finalStream = new ReadableStream({
      start(controller) {
        controller.enqueue(new TextEncoder().encode(urlDataString));
        if (response.body) {
          response.body.pipeTo(
            new WritableStream({
              write(chunk) {
                controller.enqueue(chunk);
              },
              close() {
                controller.close();
              },
              abort(err) {
                controller.error(err);
              },
            })
          );
        }
      },
    });

    return new StreamingTextResponse(finalStream);
```
    
And then the frontend code is fully custom, using my own state management and SSE handler. Will the SDK ever support extending the message schema to support custom data and custom serializers/deserializers for the text stream? Or is there a more idiomatic way that doesn't involve streaming my custom data?
    
Thanks for any help!

**Comment by danny-avila** [comment]
I have this question as well, thanks for the workaround @dnsosebee

**Comment by DeveloperRyan** [comment]
@dnsosebee I'm really struggling with this as well. I've tried a solution similar to yours with SSE on the frontend, however, I appear to be struggling to read the data properly.

I get a response similar to `data: {"id":...","object":"chat.completion.chunk","created":..,"model":"gpt-4-0613","choices":[{"index":0,"delta":{"content":"Hello"},"finish_reason":null}]}`.

Do you get a similar response, and then modify your SSE to parse this data?

**Comment by dnsosebee** [comment]
Yeah I'm not using the SDK on frontend, just using SSE much like here: https://github.com/supabase-community/nextjs-openai-doc-search/blob/3410cb787b2d663f8d5812d4a35baf881ad8754a/components/SearchDialog.tsx#L109C7-L109C12

And then when I receive the first message event, I parse it as my custom data, then parse the rest as shown in the example here

**Comment by MaxLeiter** [comment]
hey @dnsosebee sorry for the dekay. Will https://github.com/vercel-labs/ai/pull/425 sufficiently address this? Feedback appreciated!

**Comment by dnsosebee** [comment]
https://github.com/vercel-labs/ai/pull/425#issuecomment-1668716292

**Comment by MaxLeiter** [comment]
hey @dnsosebee, would you mind trying the experimental_StreamData (instructions [here](https://github.com/vercel/ai/pull/437#issuecomment-1681388183)) and reporting back? I want to confirm it meets your use-case and works as expected

**Comment by lgrammel** [comment]
You can use `StreamData`: https://sdk.vercel.ai/docs/ai-sdk-ui/streaming-data

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Feature**  
Confidence: **High**

### Rationale
This requests a new streaming-data capability rather than reporting broken or incorrect existing behavior. The subsequent introduction of `StreamData` directly addresses the requested enhancement.

### Evidence
- The issue requests SDK support for attaching custom structured data to streamed question-answer messages, including custom serialization/deserialization.
- The reporter currently relies on a custom SSE client because the SDK did not provide this capability.
- Maintainers proposed and later pointed to `experimental_StreamData`/`StreamData` as the solution.
